### DISCUSSION TITLE
Tween cancellation using CancellationToken? (added in 1.4.10)

### DISCUSSION THREAD

**Discussion by Karabaev** [KyryloKuzyk/PrimeTween#80] [Ideas]
I would like to cancel Tween/Sequence using CancellationToken. Is there any way to do that?

NOT FROM MAINTAINER: I added CancellationToken support in version 1.4.10.

**Comment by KyryloKuzyk** [comment]
Hey! There is currently no CancellationToken support in PrimeTween. Instead, PrimeTween offers a simple tween/sequence.Stop() method that can cancel animations, including those awaited with 'await'.

Can you please provide an example in which use cases you would like to use CancellationToken together with PrimeTween?

Similar threads:
https://github.com/KyryloKuzyk/PrimeTween/issues/54
https://github.com/KyryloKuzyk/PrimeTween/discussions/46

**Reply by Karabaev** [reply]
Sorry for late reply.

I have a class `AnimatedObject`, which is attached to an animated element. It has one public method `StartAnimationAsync`, which starts the animation.

Also there is a class `AnimatedObjectSpawner`, which gets an instance of the `AnimatedObject` class (from a pool), and calls its `StartAnimationAsync` method. It waits for the completion and returns the instance to the pool. The `AnimatedObjectSpawner` class subscribes to a certain external event.
When this event occurs, it is necessary to interrupt the animation and return the object with the `AnimatedObject` class to the pool.

Example methods in `AnimatedObjectSpawner`:

```c#
void OnSomeEventRaised()
{
    _cts.Cancel();    
}

async UniTask SpawnAnimatedObject()
{
    AnimatedObject instance = pool.Get();
    await instance.StartAnimationAsync(_cts.Token);
    pool.Return(instance);
}
``` 
It would be possible to add an `AnimatedObject.Interrupt` method, which would wrap `Tween.Stop`. However, to call `AnimatedObject.Interrupt`, a reference to `AnimatedObject` would need to be stored in a field.

However, by code design, it would be better to pass a `CancellationToken`, as in the example above.

Hope I managed to provide you a clear example :)

**Reply by KyryloKuzyk** [reply]
Great example, thank you!

If the only async operation that happens inside SpawnAnimatedObject() is animation, then it should be possible to do something like this. Or do you want to combine multiple async operations under one CancellationTokenSource?

```
Tween/Sequence _animation;

void OnSomeEventRaised()
{
    _animation.Stop();
}

async UniTask SpawnAnimatedObject()
{
    AnimatedObject instance = pool.Get();
    _animation = instance.StartAnimation();
    await _animation;
    pool.Return(instance);
}
```

**Reply by KyryloKuzyk** [reply]
Hey @Karabaev, I started investigating the CancellationToken topic, and the code example you shared earlier won't produce the expected result. Because cancelling the CancellationToken throws the OperationCancelledException, the `pool.Return(instance);` will never be executed in this case. The correct code should be the following:
```csharp
async UniTask SpawnAnimatedObject()
{
    AnimatedObject instance = pool.Get();
    try
    {
        await instance.StartAnimationAsync(_cts.Token);
    }
    finally
    {
        pool.Return(instance);
    }
}
```

Also, what should be the expected result of cancelling the animation via CancellationToken? The established convention in .NET is that the operation should stop doing what it was currently doing, which, in the case of the animation, means just stopping it. Or do you expect the animation to rewind to the initial state when it's cancelled via CancellationToken?

**Comment by kyubuns** [comment]
I want to use PrimeTween in combination with R3(+UniTask) (the successor to UniRx).
https://github.com/Cysharp/R3

I want to write the following code.

```csharp
var rp = new ReactiveProperty<float>(10.0f);
rp.SubscribeAwait(async (x, ct) =>
{
    await Tween.Scale(target, x, 1.0f).ToYieldInstruction().WithCancellation(cancellationToken: ct);
}, AwaitOperation.Switch);
```
However, with this, the tween will continue to run even if the CancellationToken is canceled.
I know that I can do it in the following way,

```csharp
var rp = new ReactiveProperty<float>(10.0f);
Tween tween = default;
rp.Subscribe(x =>
{
    tween.Stop();
    tween = Tween.Scale(target, x, 1.0f);
});
```

but I simply prefer the example first.

---

I understood the design concept of PrimeTween by reading #46.
I want to create a convenient extension method for my project.
But, in the code below, the class new runs every time it is called. (In Disposable.Create.)
Does anyone know a more efficient method?

```csharp
public static class PrimeTweenExtensions
{
    public static UniTask ToTask(this Tween tween, CancellationToken cancellationToken = default)
    {
        Disposable.Create(tween.Stop).AddTo(cancellationToken);
        return tween.ToYieldInstruction().ToUniTask(cancellationToken: cancellationToken);
    }
}

var rp = new ReactiveProperty<float>(10.0f);
rp.SubscribeAwait(async (x, ct) =>
{
    await Tween.Scale(target, x, 1.0f).ToTask(ct);
}, AwaitOperation.Switch);
```

**Reply by kyubuns** [reply]
This code seems better because it doesn't cause a new class to be created.
I'm not very familiar with C#. I'd like to hear the opinions of people who are familiar with it.

```csharp
public static class PrimeTweenExtensions
{
    public static UniTask ToTask(this Tween tween, CancellationToken cancellationToken = default)
    {
        new TweenDisposable(tween).AddTo(cancellationToken);
        return tween.ToYieldInstruction().ToUniTask(cancellationToken: cancellationToken);
    }

    private readonly struct TweenDisposable : IDisposable
    {
        private readonly Tween _tween;

        public TweenDisposable(Tween tween)
        {
            _tween = tween;
        }

        public void Dispose()
        {
            _tween.Stop();
        }
    }
}
```

**Reply by HikoRuwk** [reply]
@kyubuns 
Sry for Necro this. I dun know if you find solution yet. I remember DOTween plugin include a extension class for Unitask support. You can have a check to see if dat is the solution you want. 

Currently I copy that extension class and modify it to [support](https://gist.github.com/2C2C2C/551548c79e3eaa557e7f77252ece1752) PrimeTween.

**Comment by KyryloKuzyk** [comment]
Hey @Karabaev @kyubuns @HikoRuwk, I added CancellationToken support in version 1.4.10.
